metricbeat/scripts/mage/package.go

```diff
@@ -0,0 +1,233 @@
+		}
+		windowsModulesD = mage.PackageFile{
+			Mode:    0644,
+			Source:  "{{.PackageDir}}/modules.d",
+			Config:  true,
+			Modules: true,
+			Dep: func(spec mage.PackageSpec) error {
+				if err := mage.Copy(dirModulesDGenerated, spec.MustExpand("{{.PackageDir}}/modules.d")); err != nil {
+					return errors.Wrap(err, "failed to copy modules.d dir")
+				}
+
+				return mage.FindReplace(
+					spec.MustExpand("{{.PackageDir}}/modules.d/system.yml"),
+					regexp.MustCompile(`- load`), `#- load`)
+			},
+		}
+		windowsReferenceConfig = mage.PackageFile{
+			Mode:   0644,
+			Source: "{{.PackageDir}}/metricbeat.reference.yml",
+			Dep: func(spec mage.PackageSpec) error {
+				err := mage.Copy("metricbeat.reference.yml",
+					spec.MustExpand("{{.PackageDir}}/metricbeat.reference.yml"))
+				if err != nil {
+					return errors.Wrap(err, "failed to copy reference config")
+				}
+
+				return mage.FindReplace(
+					spec.MustExpand("{{.PackageDir}}/metricbeat.reference.yml"),
+					regexp.MustCompile(`- load`), `#- load`)
+			},
+		}
+	)
+
+	for _, args := range mage.Packages {
+		switch args.OS {
+		case "windows":
+			args.Spec.Files[modulesDTarget] = windowsModulesD
+			args.Spec.ReplaceFile("{{.BeatName}}.reference.yml", windowsReferenceConfig)
+		default:
+			pkgType := args.Types[0]
+			switch pkgType {
+			case mage.TarGz, mage.Zip, mage.Docker:
+				args.Spec.Files[modulesDTarget] = modulesD
+			case mage.Deb, mage.RPM, mage.DMG:
+				args.Spec.Files["/etc/{{.BeatName}}/"+modulesDTarget] = modulesD
+			default:
+				panic(errors.Errorf("unhandled package type: %v", pkgType))
+			}
+		}
+	}
+}
+
+// PrepareModulePackagingOSS generates build/package/modules and
+// build/package/modules.d directories for use in packaging.
+func PrepareModulePackagingOSS() error {
+	return prepareModulePackaging([]struct{ Src, Dst string }{
+		{mage.OSSBeatDir("modules.d"), dirModulesDGenerated},
+	}...)
+}
+
+// PrepareModulePackagingXPack generates build/package/modules and
+// build/package/modules.d directories for use in packaging.
+func PrepareModulePackagingXPack() error {
+	return prepareModulePackaging([]struct{ Src, Dst string }{
+		{mage.OSSBeatDir("modules.d"), dirModulesDGenerated},
+		{"modules.d", dirModulesDGenerated},
+	}...)
+}
+
+func prepareModulePackaging(files ...struct{ Src, Dst string }) error {
+	mg.Deps(GenerateDirModulesD)
+
+	err := mage.Clean([]string{
+		dirModulesDGenerated,
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, copyAction := range files {
+		err := (&mage.CopyTask{
+			Source:  copyAction.Src,
+			Dest:    copyAction.Dst,
+			Mode:    0644,
+			DirMode: 0755,
+		}).Execute()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+const modulesDHeader = "# Module: {{ .Module }}\n# Docs: https://www.elastic.co/guide/en/beats/{{ .BeatName }}/{{ .BeatDocBranch }}/{{ .BeatName }}-module-{{ .Module }}.html"
+
+// GenerateDirModulesD generates the modules.d directory
+func GenerateDirModulesD() error {
```

Comment:
this looks quite common to `dev-tools/mage/modules.go`, which is used by Filebeat. Would it make sense to use the same?

Reply:
Yeah, actually both could be merged, lets do it in a follow up.